### PR TITLE
Add notes on implementation of js.Publish retries

### PR DESCRIPTION
**Pull request by wallyqs**
From base implementation at  https://github.com/nats-io/nats.go/pull/930

Signed-off-by: Waldemar Quevedo <wally@synadia.com>

### CODE REVIEW

File: adr/ADR-22.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+|----------|---------------------------|
+| Date     | 2022-03-18                |
+| Author   | wallyqs                   |
+| Status   | Partially Implemented     |
+| Tags     | jetstream, client         |
+
+## Motivation
+
+When the NATS Server is running with JetStream on cluster mode, there
+can be occasional blips in leadership due which can result in a number
```
Comment (kozlovic) [inline adr/ADR-22.md:13]: typo: `leadership due which`

File: adr/ADR-22.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+### Make Publish retry as needed until context deadline
+
+In Go when using the context package, it can be possible to set the maximum deadline of the retries
+so that the client can retry as needed.  In the example below a client will attempt to publish for 10 seconds
+backing off `250ms` as needed until the service is available again:
+
+```go
+ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+defer cancel()
+_, err := js.Publish("foo", []byte("bar"), nats.Context(ctx), nats.RetryWait(250*time.Millisecond), nats.RetryAttempts(-1))
```
Comment (kozlovic) [inline adr/ADR-22.md:97]: Does that mean that this is applicable only to "context" feature?

What I mean by that is that without a context, the Publish() is still subject to the js "timeout" or "wait" timeout (that all internal APIs requests are subject to, using it as the timeout of the Request() call). So the question is: should libraries without a context still apply the same logic but use the request timeout as the upper limit?

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: adr/ADR-22.md (hunk 1 above)
Comment (wallyqs) [inline adr/ADR-22.md:13]:
```suggestion
can be occasional blips in leadership which can result in a number
```

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**